diffsync/__init__.py

```diff
@@ -692,6 +696,66 @@ def remove(self, obj: DiffSyncModel, remove_children: bool = False):
                         # Since this is "cleanup" code, log an error and continue, instead of letting the exception raise
                         self._log.error(f"Unable to remove child {child_id} of {modelname} {uid} - not found!")
 
+    def get_or_instantiate(
+        self, model: Type[DiffSyncModel], ids: Dict, attrs: Dict = None
+    ) -> Tuple[DiffSyncModel, bool]:
+        """Attempt to get the object with provided identifiers or instantiate it with provided identifiers and attrs.
+
+        Args:
+            model (DiffSyncModel): The DiffSyncModel to get or create.
+            ids (Mapping): Identifiers for the DiffSyncModel to get or create with.
+            attrs (Mapping, optional): Attributes when creating an object if it doesn't exist. Defaults to None.
+
+        Returns:
+            Tuple[DiffSyncModel, bool]: Provides the existing or new object and whether it was created or not.
+        """
+        created = False
+        try:
+            obj = self.get(model, ids)
+        except ObjectNotFound:
+            if not attrs:
+                attrs = {}
+            obj = model(**ids, **attrs)
+            # Add the object to diffsync adapter
+            self.add(obj)
+            created = True
+
+        return obj, created
+
+    def update_or_instantiate(
+        self, model: Type[DiffSyncModel], ids: Dict, attrs: Dict = None
+    ) -> Tuple[DiffSyncModel, bool]:
+        """Attempt to update an existing object with provided ids/attrs or instantiate it with provided identifiers and attrs.
+
+        Args:
+            model (DiffSyncModel): The DiffSyncModel to get or create.
+            ids (Mapping): Identifiers for the DiffSyncModel to get or create with.
+            attrs (Mapping, optional): Attributes when creating an object if it doesn't exist. Defaults to None.
+
+        Returns:
+            Tuple[DiffSyncModel, bool]: Provides the existing or new object and whether it was created or not.
+        """
+        created = False
+        try:
+            obj = self.get(model, ids)
+        except ObjectNotFound:
+            if not attrs:
+                attrs = {}
+            obj = model(**ids, **attrs)
+            # Add the object to diffsync adapter
+            self.add(obj)
+            created = True
+
+        # Update existing obj with attrs
+        if attrs:
+            existing_attrs = obj.get_attrs()
+            mutual_attrs = set(existing_attrs).intersection(attrs)
+            for attr in mutual_attrs:
+                if existing_attrs[attr] != attrs[attr]:
+                    setattr(obj, attr, attrs[attr])
```

Comment:
Is the idea here to ignore any `attrs` that aren't in the object model's base attributes? That seems odd to me. I would have expected simply:

```suggestion
        for attr, value in attrs.items():
            if getattr(obj, attr) != value:
                setattr(obj, attr, value)
```

Reply:
Do we want to consume the the `AttributeError` if the `obj` doesn't have the necessary attr and log it or raise a new custom error? Or should we default within the `getattr` to `None` if the attr doesn't exist?

Reply:
In my view this would be a coding/usage error if you're calling it with a nonexistent attribute, so I don't think we need to capture the AttributeError.